File: Daily/2444.count-subarrays-with-fixed-bounds.js
/*
 * @lc app=leetcode id=2444 lang=javascript
 *
 * [2444] Count Subarrays With Fixed Bounds
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} minK
 * @param {number} maxK
 * @return {number}
 */
var countSubarrays = function (nums, minK, maxK) {
    let invalidIndex = -1;
    let minIndex = -1;
    let maxIndex = -1;
    let subarrayNumber = 0;

    // The subarray ends at i
    for (let i = 0; i < nums.length; i++) {
        // The next subarray must starts after i
        if (nums[i] > maxK || nums[i] < minK) {
            invalidIndex = i;
            // Need to find minIndex and maxIndex again
            minIndex = -1;
            maxIndex = -1;
        }
        if (nums[i] === maxK) maxIndex = i;
        if (nums[i] === minK) minIndex = i;

        // Ensure there are minK and maxK in [invalidIndex+1, i]
        if (minIndex !== -1 && maxIndex !== -1) {
            // We can choose startIndex in [invalidIndex+1, Math.min(minIndex, maxIndex)]
            subarrayNumber +=
                Math.min(minIndex, maxIndex) - (invalidIndex + 1) + 1;
        }
    }

    return subarrayNumber;
};
// @lc code=end

File: src/2025/2444.count-subarrays-with-fixed-bounds.cpp
/*
 * @lc app=leetcode id=2444 lang=cpp
 *
 * [2444] Count Subarrays With Fixed Bounds
 */

#include <vector>

// @lc code=start
class Solution {
 public:
  long long countSubarrays(const std::vector<int>& nums,
                           const int min_k,
                           const int max_k) {
    long long subarray_count = 0;
    int current_range_start = 0;
    int current_range_min_k_index = -1;
    int current_range_max_k_index = -1;

    for (int i = 0; i < nums.size(); i++) {
      if (nums[i] < min_k || nums[i] > max_k) {
        current_range_start = i + 1;
        current_range_min_k_index = -1;
        current_range_max_k_index = -1;
      }

      if (nums[i] == min_k) {
        current_range_min_k_index = i;
      }

      if (nums[i] == max_k) {
        current_range_max_k_index = i;
      }

      if (current_range_min_k_index != -1 && current_range_max_k_index != -1) {
        subarray_count +=
            std::min(current_range_max_k_index, current_range_min_k_index) -
            current_range_start + 1;
      }
    }

    return subarray_count;
  }
};
// @lc code=end

int main() {
  Solution sol;
  sol.countSubarrays({1, 3, 5, 2, 7, 5}, 1, 5);
}
